planner/core/expression_rewriter.go

```diff
@@ -528,11 +528,18 @@ func (er *expressionRewriter) handleCompareSubquery(ctx context.Context, v *ast.
 		er.err = errors.Errorf("Unknown compare type %T", v.R)
 		return v, true
 	}
-	np, _, err := er.buildSubquery(ctx, subq, false)
+	np, hintFlags, err := er.buildSubquery(ctx, subq, handlingCompareSubquery)
 	if err != nil {
 		er.err = err
 		return v, true
 	}
+
+	noDecorrelate := hintFlags&HintFlagNoDecorrelate > 0
+	if noDecorrelate && len(extractCorColumnsBySchema4LogicalPlan(np, er.p.Schema())) == 0 {
+		er.sctx.GetSessionVars().StmtCtx.AppendWarning(ErrInternal.GenWithStack(
+			"NO_DECORRELATE() is inapplicable because there are no correlated columns."))
```

Comment:
Should we reset the `noDecorrelate` here?

Reply:
Updated.
It actually doesn't matter, I reset it anyway to be consistent with `handleExistSubquery`.